doc/release/upcoming_changes/22139.expired.rst

```diff
@@ -0,0 +1,5 @@
+* Support for Visual Studio 2015 has been removed. Please update to at least
+  Visual Studio 2019.
```

Comment:
Since we're still building with `vc141` and it is Visual Studio 2017 that defaults to that toolchain, I think that is what you want to say here. Otherwise, please make it explicit: you must use at least Visual Studio 2019 (then the first sentence should say 2017 not 2015), *and* you must use at least the `vc141` compiler toolchain.

Reply:
I think mandating a lower bound that contains a non-default toolset makes little sense. Either vs2017+vc141 or vs2019+vc142.

In the previous discussions on this, one of the last relevant constraints for moving to vc142 (as I understood it) was conda-forge still being on vc141. That's being changed now, so from that POV - and as I understood it - a move to vc142 would be possible, notwithstanding that the CI isn't at that point yet.

Reply:
> I think mandating a lower bound that contains a non-default toolset makes little sense. Either vs2017+vc141 or vs2019+vc142.

Agreed

> That's being changed now, so from that POV - and as I understood it - a move to vc142 would be possible,

It's not, `vc141` is a hard requirement until we get rid of SciPy depending on `libnpymath` at least.

Reply:
> > That's being changed now, so from that POV - and as I understood it - a move to vc142 would be possible,
> 
> It's not, `vc141` is a hard requirement until we get rid of SciPy depending on `libnpymath` at least.

If this is about the gfortran interaction, the issue with the unknown assembly section can be solved with a single flag to the same effective level of (non-)support

Reply:
Do you have a link for how to make the mingw-w64 toolchain work with vc142? I may have missed that part.

Reply:
Sure, see this [comment](https://github.com/matthew-brett/dll_investigation/issues/1#issuecomment-1100468171) and preceding discussion.

Reply:
Support for the flag  and vc142 was added in #21360. All of our windows CI jobs https://github.com/numpy/numpy/blob/555c07f71d779d59a0ea097f811442a6c1bb405c/azure-pipelines.yml#L255 and https://github.com/numpy/numpy/blob/555c07f71d779d59a0ea097f811442a6c1bb405c/.github/workflows/wheels.yml#L76-L77 run on [windows 2019 images](https://github.com/actions/runner-images/blob/main/images/win/Windows2019-Readme.md#visual-studio-enterprise-2019) with Visual Studio 2019. I would prefer not to recommend out-of-date software. I could just drop the whole sentence, recommendations are not really not part of a release note. The important part is that we are dropping `<=2015`

Reply:
Updated.

Reply:
> Support for the flag and vc142 was added in #21360.

Hmm, I just spent a lot of time doing archeology to determine exactly which versions to avoid for SciPy wheel builds because they were built with `vc142` (answer: `1.21.6`, `1.22.0` and `1.22.1`). If we now went back to `vc142` again but with a different flag, it would be great to record that somewhere prominent (e.g., the release notes, or a central doc page), because next time it'll perhaps be even harder to figure out.